cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -353,6 +353,332 @@ struct IfElseFunctor<Type, enable_if_number<Type>> {
+    out->buffers[2] = std::move(out_data_buf);
+    return Status::OK();
+  }
+
+  // ASS
+  static Status Call(KernelContext* ctx, const ArrayData& cond, const Scalar& left,
+                     const Scalar& right, ArrayData* out) {
+    const uint8_t* cond_data = cond.buffers[1]->data();
+    BitBlockCounter bit_counter(cond_data, cond.offset, cond.length);
+
+    util::string_view left_data = internal::UnboxScalar<Type>::Unbox(left);
+    int64_t left_size = left_data.size();
+
+    util::string_view right_data = internal::UnboxScalar<Type>::Unbox(right);
+    int64_t right_size = right_data.size();
+
+    // reserve an additional space
+    // todo use cond.data to calculate the out_data_buf size precisely, by summing true
+    //  bits
+    ARROW_ASSIGN_OR_RAISE(auto out_offset_buf,
+                          ctx->Allocate((cond.length + 1) * sizeof(OffsetType)));
+    auto* out_offsets = reinterpret_cast<OffsetType*>(out_offset_buf->mutable_data());
+    out_offsets[0] = 0;
+
+    // allocate data buffer conservatively
+    auto data_buff_alloc = right_size * cond.length + left_size * cond.length;
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<ResizableBuffer> out_data_buf,
+                          ctx->Allocate(data_buff_alloc));
+    uint8_t* out_data = out_data_buf->mutable_data();
+
+    int64_t offset = cond.offset;
+    OffsetType total_bytes_written = 0;
+    while (offset < cond.offset + cond.length) {
+      const BitBlockCount& block = bit_counter.NextWord();
+
+      OffsetType bytes_written = 0;
+      if (block.AllSet()) {
+        // from left
+        bytes_written = block.length * left_size;
+        for (int i = 0; i < block.length; i++) {
+          std::memcpy(out_data + i * left_size, left_data.cbegin(), left_size);
+          out_offsets[i + 1] = out_offsets[0] + (i + 1) * left_size;
+        }
+      } else if (block.NoneSet()) {
+        // from right
+        bytes_written = block.length * right_size;
+        for (int i = 0; i < block.length; i++) {
+          std::memcpy(out_data + i * right_size, right_data.cbegin(), right_size);
+          out_offsets[i + 1] = out_offsets[0] + (i + 1) * right_size;
+        }
+      } else if (block.popcount) {  // selectively copy from left
+        for (auto i = 0; i < block.length; ++i) {
+          OffsetType current_length;
+          if (BitUtil::GetBit(cond_data, offset + i)) {
+            current_length = left_size;
+            std::memcpy(out_data + bytes_written, left_data.cbegin(), current_length);
+          } else {
+            current_length = right_size;
+            std::memcpy(out_data + bytes_written, right_data.cbegin(), current_length);
+          }
+          out_offsets[i + 1] = out_offsets[i] + current_length;
+          bytes_written += current_length;
+        }
+      }
+
+      offset += block.length;
+      out_offsets += block.length;
+      out_data += bytes_written;
+      total_bytes_written += bytes_written;
```

Comment:
Since you're updating `out_data` in all these cases, you don't need to track `total_bytes_written` or `bytes_written`, it'll be just `out_data - out_data_buf->mutable_data()`.

Reply:
I need `total_bytes_written` to resize the buffer later on. But yes, I can remove one variable from here.

Reply:
What I mean is `total_bytes_written == out_data - out_data_buf->mutable_data()` since you're updating the `out_data` pointer after writing - so you don't need to track it separately, you can just compute it.